src/operator/communication/gather.rs

```diff
@@ -0,0 +1,423 @@
+                                    .value()
+                                    .clone();
+
+                                let (producer, consumer) = OneshotSpsc::new().split();
+                                let sender =
+                                    GatherProducer::new(producer, ready_callback.clone(), location);
+                                tx.send(consumer).unwrap();
+
+                                let gather_trace = if current_worker == receiver_worker {
+                                    let mut consumers = Vec::with_capacity(workers);
+                                    for _ in 0..workers {
+                                        consumers.push(rx.recv().unwrap());
+                                    }
+
+                                    self.circuit().add_exchange(
+                                        sender,
+                                        GatherConsumer::new(
+                                            consumers.into_boxed_slice(),
+                                            ready_callback,
+                                            location,
+                                        ),
+                                        self,
+                                    )
+                                } else {
+                                    self.circuit().add_exchange(
+                                        sender,
+                                        EmptyGatherConsumer::new(location),
+                                        self,
+                                    )
+                                };
+
+                                // Is `consolidate` always necessary? Some (all?) consumers may be
+                                // happy working with traces.
+                                gather_trace.consolidate()
+                            },
+                        )
+                        .clone()
+                }
+            }
+        }
+    }
+}
+
+struct GatherProducer<T> {
+    queue: Producer<T>,
+    notify: NotifyCallback,
+    location: &'static Location<'static>,
+}
+
+impl<T> GatherProducer<T> {
+    const fn new(
+        queue: Producer<T>,
+        notify: NotifyCallback,
+        location: &'static Location<'static>,
+    ) -> Self {
+        Self {
+            queue,
+            notify,
+            location,
+        }
+    }
+
+    fn notify(&self) {
+        (self.notify.get().unwrap())();
+    }
+}
+
+impl<T> Operator for GatherProducer<T>
+where
+    T: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("GatherProducer")
+    }
+
+    fn location(&self) -> OperatorLocation {
+        Some(self.location)
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<T> SinkOperator<T> for GatherProducer<T>
+where
+    T: Clone + Send + 'static,
+{
+    fn eval(&mut self, input: &T) {
+        self.queue.push(input.clone());
+        self.notify();
```

Comment:
This will spuriously invoke the callback once per sender, which doesn't affect correctness, but will cause the scheduler to do redundant work.

Reply:
Correct, I figure it's relatively cheaper than keeping track of how many values have been pushed, in most situations this will ony cause a handful of extra activations on the gathering thread